File: Bronze/2024Dec/P1.py

import sys
input = sys.stdin.readline
T = int(input())

for i in range(T):
    query = int(input())
    # Precompute valid ranges 
    ranges = []
    for length in range(2, 10):
        low = '4' * (length-1) + '5'  
        high = '4' + '9' * (length-1)
        ranges.append((int(low), int(high)))
    print(ranges)

    if query < 2:
        print(0)
        continue
        
    ans = 0
    for start, end in ranges:
        print("ans: " + str(ans))
        if start > query:
            break
        print("start,end: " + str((start,end)))
        valid_end = min(end, query)
        if valid_end >= start:
            ans += valid_end - start + 1
        print("valid_end: " + str(valid_end))
            
    print(ans)

File: Bronze/2024Dec/P1.cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
	ios::sync_with_stdio(false);
	cin.tie(NULL);

	// Read input
	int T;
	cin >> T;
	vector<int> queries(T);
	for(int &x: queries) {
		cin >> x;
	}

	// Precompute valid ranges
	vector<pair<int,int>> ranges;
	for(int len = 2; len <= 9; len++) {
		// Build lower bound (45, 445, 4445, etc)
		string low;
		for(int i = 0; i < len-1; i++) {
			low += '4';
		}
		low += '5';
		
		// Build upper bound (49, 499, 4999, etc) 
		string high = "4";
		for(int i = 0; i < len-1; i++) {
			high += '9';
		}
		
		ranges.emplace_back(stoi(low), stoi(high));
	}

	// Process each query
	for(int query: queries) {
		if(query < 2) {
			cout << "0\n";
			continue;
		}

		long long ans = 0;
		for(auto &[start, end]: ranges) {
			if(start > query) continue;
			
			int validEnd = min(end, query);
			if(validEnd >= start) {
				ans += (long long)(validEnd - start + 1);
			}
		}
		cout << ans << "\n";
	}
}